qiskit/aqua/algorithms/phase_estimators/phase_estimator.py

```diff
@@ -0,0 +1,208 @@
+
+           Raises:
+               ValueError: unless only one of `unitary` and `pe_circuit` is `None`.
+                           `num_unitary_qubits` disagrees with size of `unitary`.
+        """
+
+        # Determine if user passed a unitary, or the entire QPE circuit with the unitary
+        # already embedded, and set properties.
+        if unitary is None:
+            if pe_circuit is None:
+                raise ValueError('Only one of `unitary` and `pe_circuit` may be `None`.')
+
+            if num_unitary_qubits is None:
+                raise ValueError('`num_unitary_qubits` is required when passing `pe_circuit`.')
+            self._pe_circuit = pe_circuit
+            self._num_unitary_qubits = num_unitary_qubits
+        else:
+            if not (num_unitary_qubits is None or num_unitary_qubits == unitary.num_qubits):
+                raise ValueError('`num_unitary_qubits` disagrees with size of `unitary`.')
+            pe_circuit = PhaseEstimation(num_evaluation_qubits, unitary)
+            self._num_unitary_qubits = unitary.num_qubits
+            self._pe_circuit = pe_circuit
+
+        self._num_evaluation_qubits = num_evaluation_qubits
+
+        if state_preparation is not None:
+            self._pe_circuit = pe_circuit.compose(
+                state_preparation,
+                qubits=range(num_evaluation_qubits,
+                             num_evaluation_qubits + self._num_unitary_qubits),
+                front=True)
+        else:
+            self._pe_circuit = pe_circuit
+
+        self._measurements_added = False
+        super().__init__(quantum_instance)
+
+    def _add_classical_register(self) -> None:
+        """Explicitly add measurement instructions only if we are using a state vector simulator."""
+        if not self._quantum_instance.is_statevector and not self._measurements_added:
+            # Measure only the evaluation qubits.
+            regname = 'meas'
+            circ = self._pe_circuit
+            creg = ClassicalRegister(self._num_evaluation_qubits, regname)
+            circ.add_register(creg)
+            circ.barrier()
+            circ.measure(range(self._num_evaluation_qubits), range(self._num_evaluation_qubits))
+            self._measurements_added = True
+
+    def construct_circuit(self) -> QuantumCircuit:
+        """Return the circuit to be executed to estimate phases.
+
+        This circuit includes as sub-circuits the core phase estimation circuit,
+        with the addition of the state-preparation circuit and possibly measurement instructions.
+        """
+        self._add_classical_register()
+        return self._pe_circuit
+
+    # Favor construct_circuit
+    # @property
+    # def pe_circuit(self) -> QuantumCircuit:
+    #     """Return the phase estimation circuit. """
+    #     return self._pe_circuit
+
+    def _compute_phases(self, circuit_result: Result) -> Union[numpy.ndarray, dict]:
```

Comment:
Isn't the return type a `Counts` object in the non-statevector case?

Reply:
Yes, this is fixed.